libraries/AP_Beacon/AP_Beacon_Nooploop.cpp

```diff
@@ -0,0 +1,238 @@
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <AP_HAL/AP_HAL.h>
+#include <GCS_MAVLink/GCS.h>
+#include "AP_Beacon_Nooploop.h"
+#include <ctype.h>
+#include <stdio.h>
+
+#define NOOPLOOP_INVALID_VAL -8388000
+
+extern const AP_HAL::HAL& hal;
+
+AP_Beacon_Nooploop::AP_Beacon_Nooploop(AP_Beacon &frontend, AP_SerialManager &serial_manager) :
+    AP_Beacon_Backend(frontend)
+{
+    _uart = serial_manager.find_serial(AP_SerialManager::SerialProtocol_Beacon, 0);
+    if (_uart != nullptr) {
+        _uart->begin(serial_manager.find_baudrate(AP_SerialManager::SerialProtocol_Beacon, 0));
+    }
+}
+
+// return true if sensor is basically healthy (we are receiving data)
+bool AP_Beacon_Nooploop::healthy()
+{
+    // healthy if we have parsed a message within the past 300ms
+    return ((AP_HAL::millis() - _last_update_ms) < AP_BEACON_TIMEOUT_MS);
+}
+
+// update the state of the sensor
+void AP_Beacon_Nooploop::update(void)
+{
+    // return immediately if not serial port
+    if (_uart == nullptr) {
+        return;
+    }
+
+    // check uart for any incoming messages
+    uint32_t nbytes = _uart->available();
+    while (nbytes-- > 0) {
+        uint8_t b = _uart->read();
+        MsgType type = parse_byte(b);
+        if (type == MsgType::NODE_FRAME2) {
+            if (_anchor_pos_avail) {
+                parse_node_frame2();
+            } else {
+                request_setting();
+            }
+        } else if (type == MsgType::SETTING_FRAME0) {
+            parse_setting_frame0();
+        } 
+    }
+}
+
+void AP_Beacon_Nooploop::request_setting()
+{
+    if (_uart->txspace() >= 128) {
+        uint8_t buf[128] = {0x54, 0, 1}; 
```

Comment:
That's a very large chunk of space to have on the stack.  I wonder if we can do better....

Also, removing the `=` here would be good.  If the compiler isn't doing a good job then you end up constructing an object on the stack and then copying it into the LHS....

Reply:
Maybe I can change it to
`_uart->write(0x54);`
`_uart->write(0);`
`_uart->write(1);`
`for (int i=0;i<124;i++) _uart->write(0);`
`_uart->write(0x55);`
not sure if it is better this way